# Feel free to reach out to us directly!

### [Rocio Martinez-Nunez](https://twitter.com/rociotmartinez)
- RNA-seq and RNA things:
  - contact: [@rociotmartinez](https://twitter.com/rociotmartinez); rocio.martinez_nunez@kcl.ac.uk  

---

### [Mike Lee](https://astrobiomike.github.io)  
- I focus primarily on microbial ecology and evolution, and have a good amount of experience with amplicon/marker-gene analysis, metagenomics, metatranscriptomics, bash, R, and python
  - contact: [@AstrobioMike](https://twitter.com/AstrobioMike); Mike.Lee@nasa.gov

---

### Rachel Lombardi
-  I focus on RNAseq to investigate the plant's response to pathogen infections.
  - contact: rllombardi@ucdavis.edu

---

### [Sichong Peng](http://sichong.site/)
- My study focuses on completing functional annotation of the horse genome. I have a strong interest of applying genomics tools to non-model organisms to improve health care of companion and agricultral animals.
  - contact: scpeng@ucdavis.edu

---

### [Chissa Rivaldi](https://twitter.com/powerofcheez)
- I am rooted in population/landscape genetics, parasite, and amplicon sequencing studies. I script in bash, python, and R, and increasingly automation flows like snakemake. 
  - contact: [@powerofcheez](https://twitter.com/powerofcheez); clrivaldi@gmail.com 

---

### [Marian L. Schmidt](https://marschmi.github.io/marschmi.github.io/)  
- I study the drivers of aquatic bacterial community composition & the evolutionary controls on bacterial genome size. I have experience (& am happy to help others!) with 16S rRNA amplicon data sets, recovering genomes from complex metagenomic sequencing data sets, using version control with git/GitHub, and programming in R (Yay Rmarkdown!) & bash. 
  - contact: [@micro_marian](https://twitter.com/micro_marian); marschmi "at" utexas "dot" edu
---
